Ability to add 0K elastic scattering from an ENDF file

paulromano opened the pull request:
This pull request adds the ability to add 0K elastic scattering to an `IncidentNeutron` instance from an ENDF file that can be used for DBRC (or equivalent). It uses the resonance reconstruction features in openmc.data to generate a 0K pointwise cross section. Two new functions in openmc.data were added, `linearize` and `thin`, the former of which will come up with a reasonable set of (x,y) points for a function and the latter of which will then "thin" out that set of points. The nuclear data format didn't need any changes. Some extra logic was required so that the 0K is written/read from an HDF5 file if present.

I've updated our NNDC data to include 0K elastic scattering data for U238, U235, and Pu239. Now the `test_resonance_scattering` test will actually run DBRC (and ARES, the fancy method by @walshjon) for real, not just faking it with 293 K data.

This PR fulfills a promise I made to @bforget to take a look at our heavy nuclide resonance scattering capabilities and make sure they were in a usable state.

smharper commented:
Nice addition!  I'll try to rerun the Mosteller benchmark for verification.

smharper commented:
Here's my initial feedback:

1. It looks like the process of starting with 1000 equal lethargy points and refining from there based off of the error at midpoints is not precise enough to capture all of the resonances.  For example, it completely misses big resonances at 5355 and 5367 keV and a little guy at 89.25 eV for U-238.  Maybe we should try to use the resonance locations stored in the ENDF file to make a more targeted initial grid.

2. I'm a little worried about the number of points in general.  The 0.1 K data from the MCNP ENDF-7.1 U-238 has 14760 points between 1 and 400 eV.  The code in this PR generates 7114.

3. If we just want this feature for DBRC, then maybe we should only linearize the data up to 400 eV rather than the full RRR.  This gives us fewer cross sections to worry about catching, and it might help with the runtime as well.

4. I'd prefer that this method not overwrite existing 0 K or 0.1 K data.

5. The `kTs` array isn't updated with the new 0 K entry.

smharper commented:
To clarify point number 4, I don't think we need to have `nuclide[2].xs['0.0K']` and `nuclide[2].xs['0.1K']`, I just want to be warned before existing 0.1 K or 0 K data is overwritten.  Maybe the user has to pass a flag like `overwrite_existing = True`.

paulromano commented:
Thanks for the quick yet thoughtful review, @smharper. Really good point on using the resonance energies as starting points -- I'll add that in.

As for the number of points, I'm not sure how tight the MCNP library maintainers chose their tolerances when generating data, but it might well be tighter than the default I'm using (0.1%). I'll do some comparisons of the MCNP library to what I'm generating to see how bad ours looks.

Good point on only going up to 400 eV. I'll look back at the literature to see if that is justified (feel free to tell me what you're basing that number off of if you have something in mind).

For the `kTs` array, that is actually done by design. In my view, `kTs` really tells you "expect cross sections for all reactions at these temperatures". If we added 0 to the `kTs` array, then it will expect to see other reactions at 0 K as well. Instead, the 0 K energy grid and elastic scattering xs appear behind the scenes, and when they go into the Fortran side, there are special arrays on the `Nuclide` derived type for them.

walshjon commented:
It will be nice to have this reconstruction.  It was always a little hacky the way we had to generate 0 K ACE data and specify it in `settings.xml` for the resonance scattering treatments.

@smharper is right on point #1: resonance reconstruction should include the resonance parameter energies.

400 eV should be fine as a cutoff for virtually all power reactor problems (I believe MC21 uses 1 keV).  I'd prefer that we reconstruct the data for the entire RRR (in case some users want it) but set a default at 400 or 1000 eV to address the runtime concern.

paulromano commented:
@walshjon You're saying we should set the default max E for DBRC to 400 or 1000 eV?

walshjon commented:
@paulromano Yes, anything around those energies should be high enough to preserve accuracy and low enough to not compromise runtime too much.  An option to set the cutoff higher would be nice, also.

paulromano commented:
So right now, there really are no 'defaults' for resonance scattering -- the user has to specify everything. I think we should have some sensible defaults; to that end, I'd like to propose the following changes:
1. Get rid of the `ResonanceScattering` class
2. Have `Settings.resonance_scattering` take a dictionary with keys `method`, `energy_min`, `energy_max`, `nuclides`, and `enable`.
3. Have the method and min/max energies be applied to all nuclides (@walshjon is there any reason why one would want to apply different methods/energy ranges to different nuclides?)
4. Make the default method 'ares', default minimum energy 1e-5 eV, and default maximum energy 1 keV.
5. If `enable` is set to `True` but no `nuclides` are specified, apply the chosen or default method to all nuclides with available data.

So, specifying parameters would look something like:
```Python
settings = openmc.Settings()
settings.resonance_scattering = {
  'enable': True,
  'method': 'dbrc',
  'energy_max': 400,
  'nuclides': ['U238', 'U235', 'U234']
}
```
In the simplest case that someone has an appropriate library and they just want to account for resonance scattering, they would simply set:
```Python
settings.resonance_scattering = {'enable': True}
```
What say ye, @walshjon and @smharper?

paulromano commented:
Also, since cxs (constant cross section) is the default when not using a resonance elastic scattering method, it doesn't really make sense to give that as an option (it would be equivalent to just not listing a nuclide in `nuclides`). So, I'll also propose removing that option.

walshjon commented:
I had coded a 1 keV E_max in the original resonance scattering implementation.  That got lost between release 0.7.1 and now.

I don't believe there's a compelling reason to have different methods/cutoffs for different nuclides given the add complications.  You could imagine some nuclides having different energy ranges over which resonance scattering is important, but I think a single energy is the may to go.

The rest of your proposal looks good to me @paulromano .

The constant cross section method had been an option in order to allow the altering of its min/max energies in a manner consistent with DBRC and ARES, but given that it should not be applied in order to model resonance scattering, I agree with removing it.

paulromano commented:
Awesome. Also, I realized after writing it that we probably don't need a ridiculously low minimum energy as I proposed. Do you think 1 eV is a good default minimum?

paulromano commented:
Oh, sorry, I guess I just missed it -- the defaults you put in [are still there](https://github.com/mit-crpg/openmc/blob/021e94060e8eae1abddb56cf7ea24658145d0a33/src/nuclide_header.F90#L117-L118). I'll adopt those values.

walshjon commented:
I think those are probably still good.  I took them from TRIPOLI and MC21.  The 0.01 eV min is needed for some Pu resonances, at least.

paulromano commented:
FYI Not really related to resonance elastic scattering, but I just threw in a commit that fixes reading ACE files that have numbers less than 1e-100 present.

smharper commented:
RE 400 eV: I picked that number because I assumed the code never sampled target velocities above that energy, but I see now that it will indeed sample above 400 eV if you give it the 0K data.  I've never thought about it before, but I suppose resonant scattering can still be important at very high energies.

RE global max/min E: Maybe a user would want to vary the energy range across nuclides for runtime purposes.  For example, maybe you cover the entire RRR for a big absorber like U-238 in an LWR but only the low-lying resonances for your less important nuclides.

walshjon commented:
It's 400**kT** that is a common cutoff for free gas target velocity sampling.  For power reactors, treating nuclei as stationary targets will be fine for neutrons in the keV range and above.  The hotter the system, the less true this becomes.  So really, tying the max to temperature may be a good idea.  It just has to be higher than 400kT, which is too low.

@viviantran performed a systematic investigation of applying resonance scattering to specific nuclides over varying energy ranges for a depleted LWR pin cell.  We found that ~100's eV was sufficient (and often overkill) for all nuclides (to get the correct k_eff) and there wasn't a huge performance benefit to paring back the E_max value on a nuclide-specific basis.

smharper commented:
Ah, well 400 eV was just a goofy number all around then.  Perhaps 100's eV is sufficient for reactor k_eff, but I wonder if we'll run into troubles getting resonance range absorption cross sections correct.

paulromano commented:
I've made all the changes that were discussed and believe I have addressed all comments. @walshjon @smharper Let me know if you would like to see any other changes.

walshjon commented on src/input_xml.F90:
```diff
@@ -851,59 +849,53 @@ subroutine read_settings_xml()
+      if (check_for_node(node_res_scat, "energy_max")) then
+        call get_node_value(node_res_scat, "energy_max", res_scat_energy_max)
+      end if
+      if (res_scat_energy_max < ZERO) then
```
less than `res_scat_energy_min`

walshjon commented on src/physics.F90:
```diff
@@ -828,27 +828,28 @@ subroutine sample_target_velocity(nuc, v_target, E, uvw, v_neut, wgt, xs_eff, kT
 
       ! upper resonance scattering energy bound (target is at rest above this E)
-      if (E > nuc % E_max) then
+      if (E > res_scat_energy_max) then
```
Here, or somewhere earlier on, we should also make sure that resonance scattering is not done in the URR by either checking that E < E_urr_min (here) or res_scat_energy_max < E_urr_min (earlier on)

smharper commented:
We're still missing resonances.  Take a look at 3963 keV, for example.  I'm seeing a 30% error in that scattering minimum.  I wonder if we can analytically compute or approximate the locations of the minima and maxima associated with each scattering resonance?

I'm inclined to think that it might be faster (because of fewer memory reallocs with pop and append) and more accurate to simply start with a deluge of 100,000 or maybe even 1,000,000 energy points and do aggressive thinning from there.  You might be able to make the thinning operation really fast with clever use of numpy.  I'm thinking of some iterative algorithm where you use a vectorized array operation to get an array of `True` `False`.  If I'm not mistaken, indexing the original array with this boolean array will give you a "view" of the original array so you can do many thinning iterations without reallocating memory.

Also, now that we have the capability to make our own 0K data, I think we should be pushing resonant scattering harder.  Perhaps turn it on by default and raise a fatal error if it's on but there is no 0K data.

walshjon commented:
Good point on the scattering dips @smharper .  I think that could be a tough problem to solve with bisecting energies, depending on the shape of the dip.  However, I don't think we're doing much/any worse than codes that also have a resonance integral thinning tolerance (which removes points near the minima due to their low contribution).

I'm good with this PR pending the addressing of the recent comments and maybe a couple of results showing that we're still getting good answers (for k_eff, scattering kernels, or both)

smharper commented:
I asked @cjosey about the analytic optima finding, and he says something about it being "difficult if not intractable" so I reckon that's a no-go.

cjosey commented:
@smharper asked me to look into this a bit further.  How NJOY does it (https://t2.lanl.gov/nis/publications/NJOY2012.pdf, pg. 46) is to start with an energy grid with the peaks and all half-height energies.  These only need to be known approximately, so you can probably get away with E, E +/- Gamma/2 for all  Gamma.  I don't remember off the top of my head whether or not Gamma as given in ENDF is "actual width" or some derivative.

A second alternative is to somewhat linearize the problem, and use SLBW with one resonance.  Since the function is analytic, one can compute exact peaks and half-widths, ignoring channel-channel interference.

paulromano commented:
Adding the half-height energies definitely seems to have helped. At the very least, I checked that 3963 eV resonance in U238 and now it looks like:
![3963](https://cloud.githubusercontent.com/assets/743095/24311023/ad638612-10a8-11e7-8d4f-c25f4aa97077.png)

paulromano commented:
Something to keep in mind -- the MCNP data is actually at 0.1 K, not 0 K, so we shouldn't expect it to match exactly. When I first wrote the resonance reconstruction algorithm in `openmc.data`, I did extensive comparison with files produced by NJOY 2012 that were at exactly 0 K and the comparisons looked great (see #721 for details -- there's a link to a tarball with plots for all nuclides). For elastic scattering in U238, the relative differences were less than 6e-6. To get things matching that well, I also had to fudge the physical constants in OpenMC to match what NJOY was using (since theirs were old).

walshjon commented:
I remember being curious: Did you have to use the same energy grid as the ACE data to get that agreement?  Or set incredibly tight tolerances for both codes?  Or did it really just come out that good using two different reconstruction algorithms set to 0.1%?

paulromano commented:
To get agreement that good, yes, the energy points do need to be the same (in this case it was a PENDF rather than an ACE file). When compared on different grids, of course the errors become higher as you'd expect. The biggest thing I was concerned about though was making sure that at a given energy we were getting the correct value.

paulromano commented:
Just pushed some commits that add the suggested checks on energy by @walshjon and will now do reconstruction starting with an energy grid that includes the resonance peak energies as well as E +/- Gamma/2.

smharper commented on openmc/settings.py:
```diff
@@ -778,10 +787,27 @@ def dd_count_interactions(self, interactions):
+                               _RES_SCAT_METHODS)
+            elif key == 'energy_min':
+                name = 'resonance scattering minimum energy'
+                cv.check_type(name, value, float)
```
Change `float` to `(float, int)` here and below.

paulromano replied:
I'll use `Real` which covers both, thanks for pointing that out

smharper commented on src/input_xml.F90:
```diff
@@ -5280,79 +5271,51 @@ subroutine assign_temperatures(material_temps)
+        if (nuc % urr_present) then
+          if (res_scat_energy_max > nuc % urr_data(1) % energy(1)) then
+            call fatal_error("Resonance scattering maximum energy is above the &
+                 &bottom of the unresolved resonance region.")
```
This can cause a problem if you have multiple resonant scattering isotopes.  Maybe a user wants to resolve the resonance scattering of U-238 up to the end of its RRR at 20 keV but this will error on the U-235 URR which starts at 2.25 keV.  (Removing this fatal error also makes it easy to just tell the code "use resonant scattering at every energy you can" by e.g. setting the max energy to 20 MeV rather than having to look-up URR energies.)

paulromano replied:
Good point. It looks like resonant scattering [won't be used in the URR](https://github.com/mit-crpg/openmc/blob/4353e57f956de3b4baba77e38d50ddd49668efc2/src/physics.F90#L465-L470) anyway.

smharper commented:
I've run into a bit of a stumbling point trying to validate this PR.  I'm getting a rare (took 2.6 million neutrons to find it) segfault in `calculate_nuclide_xs` of cross_section.F90.  I tried going to debug mode, but that's uncovered a different (possibly unrelated?) error that looks like it has nothing to do with this PR but with our converted MCNP ENDF/B-VII.1 data.  With debug mode on the 900 K O16 data causes an error at line 517 inside `nuclide_create_derived` of nuclide_header.F90.  For some unknown reason `this % sum_xs(t) % elastic` has a length of 2941 but `rx % xs(t) % value` is 2933.

I also have no idea how long this bug has been around because I don't know that I've ever run the MCNP ENDF7.1 900 K O16 data in debug mode before.

smharper commented:
Oh, actually it looks like somehow the wires are getting crossed between the 600 K and 900 K O16 data.  600 K has 2933 energy grid points and 900 K has 2941

smharper commented on src/nuclide_header.F90:
```diff
@@ -309,24 +293,35 @@ subroutine nuclide_from_hdf5(this, group_id, temperature, method, tolerance, &
+    ! Read energy grid
+    energy_group = open_group(group_id, 'energy')
+    do i = 1, n_temperature
       energy_dset = open_dataset(energy_group, temp_str)
```
Add
```fortran
      temp_str = trim(to_str(temps_to_read % data(i))) // "K"
```

paulromano replied:
Nice catch, thanks!

smharper commented on src/nuclide_header.F90:
```diff
@@ -309,24 +293,35 @@ subroutine nuclide_from_hdf5(this, group_id, temperature, method, tolerance, &
-    call close_group(kT_group)
+    ! Check for 0K energy grid
+    if (object_exists(energy_group, '0K')) then
+      energy_dset = open_dataset(energy_group, temp_str)
```
Change `temp_str` to `'0K'`

smharper commented:
The energy grid produced by this method is surprisingly sparse at low energies.  For U-238, the first two grid points are 1e-5 and 1.56 eV.  This sparse grid has serendipitously unearthed a small bug in `sample_target_velocity`.

If a neutron is significantly below 1.56 eV, you'll find that both `i_E_low` and `i_E_up` will evaluate to 1.  It looks like both DBRC and ARES will try to evaluate an array at `i_E_up - 1` which will raise a runtime error if you're running in debug mode.  For example:
```
At line 978 of file /home/smharper/openmc/src/physics.F90
Fortran runtime error: Index '0' of dimension 1 of array 'nuc%xs_cdf' below lower bound of 1
```

paulromano commented:
@smharper Thanks for such a careful review, and for uncovering that bug. It may seem a bit odd that there are no points in U238 between 1e-5 and 1.5 eV, but if you look at the cross section on a lin-lin plot, you'll see why:
![u238_mt2](https://cloud.githubusercontent.com/assets/743095/24354787/3c5f7b36-12c1-11e7-8f5e-8ff91e9cdfee.png)
The cross section is nearly linear, so when you thin out points with a 0.1% error tolerance, all of those points get removed.

I'm a bit confused on how to fix the bug. There are a few potential places where the current logic is prone to errors:
- [here](https://github.com/mit-crpg/openmc/blob/4353e57f956de3b4baba77e38d50ddd49668efc2/src/physics.F90#L918-L919) -- this can be fixed with a check for `i_E_up > 1`
- [here](https://github.com/mit-crpg/openmc/blob/4353e57f956de3b4baba77e38d50ddd49668efc2/src/physics.F90#L976-L979) -- this can be fixed with a check similar to what is done for `i_E_low` immediately above it (that being said, I'm confused why when `m` is calculated the numerator is done on indices between `i_E_up - 1` and `i_E_up` whereas the denominator is `i_E_up` to `i_E_up + 1`)
- and [here](https://github.com/mit-crpg/openmc/blob/4353e57f956de3b4baba77e38d50ddd49668efc2/src/physics.F90#L1001-L1008) -- this one I'm really not sure about. Any suggestions @walshjon?

walshjon commented:
> here -- this can be fixed with a check for i_E_up > 1

If E_low and E_up both fall between the same energy grid points, the maxval() portion should not be performed.  Also, the maxval() call should check i_E_low + 1 : i_E_up (instead of i_E_low + 1 : i_E_up - 1)

> here -- this can be fixed with a check similar to what is done for i_E_low immediately above it (that being said, I'm confused why when m is calculated the numerator is done on indices between i_E_up - 1 and i_E_up whereas the denominator is i_E_up to i_E_up + 1)

From a 09/10/2014 comment on the #256 PR thread:
> I believe those indices are correct. They are the bounding indices of E_low. The xs_cdf indices are each one less because xs_cdf value i actually contains the integrated CDF value up to i+1.

> and here -- this one I'm really not sure about. Any suggestions @walshjon?

The logic and calcs are pretty dependent on lower and upper values not being bounded by the same two grid points.  I'm inclined to default to constant cross section, free gas in such situations since they are an indication of smoothly varying cross section over the range of attainable relative energies.  (this would apply to all three bullets)

paulromano commented:
@walshjon I agree that when energies are bounded by the same grid points, reverting to CXS makes sense. I implemented that but I'm still running into a problem with [this line](https://github.com/mit-crpg/openmc/blob/4353e57f956de3b4baba77e38d50ddd49668efc2/src/physics.F90#L1001), namely because `i_E_low - 1` can still be zero if `i_E_low` is 1. Any suggestion for that one?

walshjon commented:
Man, whoever coded these indices was a real jerk!  Or else just took a dense energy grid for granted...

It looks like that index offset for the xs_cdf throws off [1005](https://github.com/mit-crpg/openmc/blob/4353e57f956de3b4baba77e38d50ddd49668efc2/src/physics.F90#L1005) and [1008](https://github.com/mit-crpg/openmc/blob/4353e57f956de3b4baba77e38d50ddd49668efc2/src/physics.F90#L1008) also.

How about:
>
          cdf_rel = cdf_low + prn() * (cdf_up - cdf_low)
          if (cdf_rel < nuc % xs_cdf(1) then
            i_E_rel = 1
            E_rel = nuc % energy_0K(i_E_low + i_E_rel - 1)
            m = (nuc % xs_cdf(i_E_low + i_E_rel - 1) - ZERO)
                 & / (nuc % energy_0K(i_E_low + i_E_rel) &
                 & -  nuc % energy_0K(i_E_low + i_E_rel - 1))
            E_rel = E_rel + (cdf_rel - ZERO) / m
          else if (i_E_low == 1) then
            i_E_rel = 1 + binary_search(nuc % xs_cdf(i_E_low:i_E_up), &
                 & i_E_up - i_E_low + 1, cdf_rel)
            E_rel = nuc % energy_0K(i_E_low + i_E_rel - 1)
            m = (nuc % xs_cdf(i_E_low + i_E_rel - 1) &
                 & - nuc % xs_cdf(i_E_low + i_E_rel - 2)) &
                 & / (nuc % energy_0K(i_E_low + i_E_rel) &
                 & -  nuc % energy_0K(i_E_low + i_E_rel - 1))
            E_rel = E_rel + (cdf_rel - nuc % xs_cdf(i_E_low + i_E_rel - 2)) / m
          else
            i_E_rel = binary_search(nuc % xs_cdf(i_E_low-1:i_E_up), &
                 & i_E_up - i_E_low + 2, cdf_rel)
            E_rel = nuc % energy_0K(i_E_low + i_E_rel - 1)
            m = (nuc % xs_cdf(i_E_low + i_E_rel - 1) &
                 & - nuc % xs_cdf(i_E_low + i_E_rel - 2)) &
                 & / (nuc % energy_0K(i_E_low + i_E_rel) &
                 & -  nuc % energy_0K(i_E_low + i_E_rel - 1))
            E_rel = E_rel + (cdf_rel - nuc % xs_cdf(i_E_low + i_E_rel - 2)) / m
          end if

paulromano commented:
@walshjon Would this work too?
```diff
diff --cc src/input_xml.F90
index c42820b,c42820b..9bf7832
--- a/src/input_xml.F90
+++ b/src/input_xml.F90
@@@ -5287,7 -5287,7 +5287,8 @@@ contain
  
          ! Build CDF for 0K elastic scattering
          xs_cdf_sum = ZERO
--        allocate(nuc % xs_cdf(size(nuc % energy_0K)))
++        allocate(nuc % xs_cdf(0:size(nuc % energy_0K)))
++        nuc % xs_cdf(0) = ZERO
  
          associate (E => nuc % energy_0K, xs => nuc % elastic_0K)
            do j = 1, size(E) - 1
```

walshjon commented:
I think you'd need to couple that with the following changes ([instead of this](https://github.com/mit-crpg/openmc/blob/4353e57f956de3b4baba77e38d50ddd49668efc2/src/physics.F90#L1001-L1008)), plus a close look at anywhere else the xs_cdf array is accessed.

I don't remember why I settled on `xs_cdf(i)` value including the integral from `i` to `i+1`, but I wouldn't be surprised if it was something subtle that will quietly break in changing to the more intuitive indexing scheme you have.  In fact, I remember starting out with your scheme.

        i_E_rel = binary_search(nuc % xs_cdf(i_E_low:i_E_up), &
             & i_E_up - i_E_low + 1, cdf_rel)
        E_rel = nuc % energy_0K(i_E_low + i_E_rel - 1)
        m = (nuc % xs_cdf(i_E_low + i_E_rel) &
             & - nuc % xs_cdf(i_E_low + i_E_rel - 1)) &
             & / (nuc % energy_0K(i_E_low + i_E_rel) &
             & -  nuc % energy_0K(i_E_low + i_E_rel - 1))
        E_rel = E_rel + (cdf_rel - nuc % xs_cdf(i_E_low + i_E_rel - 1)) / m

paulromano commented:
Are you sure? The change I proposed doesn't actually change the indexing, it just adds one extra point to the xs_cdf array (at index 0). So, whatever value was at `xs_cdf(1)` before would still be at `xs_cdf(1)`. It's just that now `xs_cdf(0)` would give you back zero instead of being out of bounds.

walshjon commented:
I misread your change and the changes I suggested would be required if you shifted the xs_cdf array and added `ZERO` at `xs_cdf(1)`.

Wouldn't your actual solution give a problem at [here](https://github.com/mit-crpg/openmc/blob/4353e57f956de3b4baba77e38d50ddd49668efc2/src/physics.F90#L1003) in the case that both `i_E_low = 1` and `i_E_rel = 0`?

paulromano commented:
Well, `binary_search` always returns an index >= 1, so `i_E_rel` shouldn't ever be zero. The only place `i_E_low + i_E_rel - 2` (which could be zero) is used is as an index to `xs_cdf`.

walshjon commented:
After another look, I'm still not crazy about relying on the re-indexing of `xs_cdf` in `binary_search` but I think your solution is the best option @paulromano .

paulromano commented:
I just pushed some commits which should take care of the outstanding issues:
- To @smharper's concern: I've adopted a method from Fudge for reconstructing resonances. Instead of just taking the resonance peak energy and two points on the side (corresponding to +/- gamma/2), it now takes the peak +/- gamma*w for 61 values of w (that are log-spcaed between 1e-3 and 1e3 -- 10 points per decade). When I compare the resulting cross section to one with 500,000 points, the maximum error I see for U238 is about 0.3%, and the vast majority of the points are below 0.1%.
- When `i_E_low == i_E_up`, ARES will now revert to CXS.
- I've extended the CDF for ARES to zero to take care of the indexing issues.

Reconstructing resonances is a bit slow now since there are many more points it needs to evaluate, but I'm finding that the time is actually pretty comparable to NJOY/RECONR. There are definitely ways we can improve the performance (most notably, there is a 3x3 complex matrix inversion for reconstructing Reich-Moore that could be improved -- at the moment we rely on np.linalg.inv). But, correctness is more important than performance, and I'm pretty confident now that our 0 K data is correct. @smharper hopefully you agree when looking over the new data this time around.

@walshjon and I have been discussing further improvements to ARES, so there may be a separate PR on that in the future. One improvement I was able to make as part of this PR is in calculating a value for Maxwellian rejection. Before this required two `sqrt` and `exp` evaluations per iteration -- I was able to simplify the expression so that it only requires a single `exp`. @walshjon I'd appreciate if you can double check my math and make sure what I have makes sense.

nelsonag commented:
Should there be some form of an example notebook for this?

On Thu, Mar 30, 2017 at 11:34 AM, Paul Romano <notifications@github.com>
wrote:

> I just pushed some commits which should take care of the outstanding
> issues:
>
>    - To @smharper <https://github.com/smharper>'s concern: I've adopted a
>    method from Fudge for reconstructing resonances. Instead of just taking the
>    resonance peak energy and two points on the side (corresponding to +/-
>    gamma/2), it now takes the peak +/- gamma*w for 61 values of w (that are
>    log-spcaed between 1e-3 and 1e3 -- 10 points per decade). When I compare
>    the resulting cross section to one with 500,000 points, the maximum error I
>    see for U238 is about 0.3%, and the vast majority of the points are below
>    0.1%.
>    - When i_E_low == i_E_up, ARES will now revert to CXS.
>    - I've extended the CDF for ARES to zero to take care of the indexing
>    issues.
>
> Reconstructing resonances is a bit slow now since there are many more
> points it needs to evaluate, but I'm finding that the time is actually
> pretty comparable to NJOY/RECONR. There are definitely ways we can improve
> the performance (most notably, there is a 3x3 complex matrix inversion for
> reconstructing Reich-Moore that could be improved -- at the moment we rely
> on np.linalg.inv). But, correctness is more important than performance, and
> I'm pretty confident now that our 0 K data is correct. @smharper
> <https://github.com/smharper> hopefully you agree when looking over the
> new data this time around.
>
> @walshjon <https://github.com/walshjon> and I have been discussing
> further improvements to ARES, so there may be a separate PR on that in the
> future. One improvement I was able to make as part of this PR is in
> calculating a value for Maxwellian rejection. Before this required two
> sqrt and exp evaluations per iteration -- I was able to simplify the
> expression so that it only requires a single exp. @walshjon
> <https://github.com/walshjon> I'd appreciate if you can double check my
> math and make sure what I have makes sense.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mit-crpg/openmc/pull/840#issuecomment-290449285>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA_TM2qhHWwzmfJmOo1iBaTmrevpKGlFks5rq8ubgaJpZM4MkUdh>
> .
>

paulromano commented:
@nelsonag I'm planning another PR that adds a few notebooks. I can update the nuclear data notebook as part of that PR to show this capability.

smharper commented:
Looking good so far.  I don't see any errors above 0.5% in the 238 grid, and I can run 100,000 neutrons in debug mode without any errors.  I'm running the Mosteller benchmark now.  I'll get back to you when that's done.

smharper commented:
Here are some preliminary results for the UO2 enrichment series of Mosteller.  It looks good so far, but I'm waiting on a simulation with 10x more neutrons before I feel decided.  The line labeled "NJOY" is the MCNP 0.1 K data.

![image](https://cloud.githubusercontent.com/assets/4622797/24559194/ccd3d520-160b-11e7-9750-b80abf68513e.png)

Any thoughts on what I said earlier about turning on resonant scattering by default?

paulromano commented:
@smharper Yes, my thought is that we definitely should. I'm of the opinion that the default behavior should be for the code to be correct. That being said, @walshjon and I found a few bugs and potential improvements in the ARES algorithm, so I'd like to get that fixed before we make it default.

smharper commented:
My bigger Mosteller run was terminated prematurely.  I'll restart it and post the results here, but in the meantime, I am confident enough to merge this PR.  Thanks, @paulromano!
